summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala

```diff
@@ -1,102 +1,254 @@
+    case (shardId, (key, batchId), (timestamp, value)) => (timestamp, (key, value))
+  }
+
+  def shardedToSummerInput[K, V]: (Sharded[K, V] => SummerInput[K, V]) = {
+    case (shardId, aggregateKey, aggregateValue) => Some((aggregateKey, aggregateValue))
+  }
+
+  private case class OutgoingSummersProps(
+    allSummers: Boolean,
+    batcher: Batcher,
+    semigroup: Semigroup[_],
+    shards: KeyValueShards
+  )
 }
 
 /**
  * This class encapsulates logic how to build `StormTopology` from DAG of the job, jobId and options.
  */
-case class StormTopologyBuilder(options: Map[String, Options], jobId: JobId, stormDag: Dag[Storm]) {
+private[storm] case class StormTopologyBuilder(options: Map[String, Options], jobId: JobId, stormDag: Dag[Storm]) {
   import StormTopologyBuilder._
 
-  def build: StormTopology = registerEdges(registerNodes).build(jobId)
-
-  private def registerNodes: Topology =
-    stormDag.nodes.foldLeft(Topology.empty) {
-      case (topology, node: SummerNode[Storm]) =>
-        topology.withBolt(getNodeName(node), SummerBoltProvider(this, node).apply)._2
-      case (topology, node: FlatMapNode[Storm]) =>
-        topology.withBolt(getNodeName(node), FlatMapBoltProvider(this, node).apply)._2
-      case (topology, node: SourceNode[Storm]) =>
-        topology.withSpout(getNodeName(node), SpoutProvider(this, node).apply)._2
-    }
-
-  private def registerEdges(topologyWithNodes: Topology): Topology =
-    stormDag.nodes.foldLeft(topologyWithNodes) {
-      case (topology, node: SummerNode[Storm]) =>
-        val edgeType = EdgeType.AggregatedKeyValues[Any, Any](getSummerKeyValueShards(node))
-        registerIncomingEdges(topology, node, edgeType)
-      case (topology, node: FlatMapNode[Storm]) =>
-        registerIncomingEdges(topology, node, flatMapEdgeType(node))
-      case (topology, node: SourceNode[Storm]) => topology
-        // ignore, no incoming edges into sources
+  def build: StormTopology = {
+    val topology = stormDag.nodes.foldLeft(Topology.empty) {
+      case (currentTopology, node: SummerNode[Storm]) =>
+        register(currentTopology, node, SummerBoltProvider(this, node))
+      case (currentTopology, node: FlatMapNode[Storm]) =>
+        register(currentTopology, node, FlatMapBoltProvider(this, node))
+      case (currentTopology, node: SourceNode[Storm]) =>
+        register(currentTopology, node, SpoutProvider(this, node))
     }
+    topology.build(jobId)
+  }
 
-  private def flatMapEdgeType(node: FlatMapNode[Storm]): EdgeType[_] = {
-    val usePreferLocalDependency = getOrElse(node, DEFAULT_FM_PREFER_LOCAL_DEPENDENCY)
-    logger.info(s"[${getNodeName(node)}] usePreferLocalDependency: ${usePreferLocalDependency.get}")
-    if (usePreferLocalDependency.get) {
-      EdgeType.itemWithLocalOrShuffleGrouping[Any]
+  private def register(topology: Topology, node: StormNode, provider: ComponentProvider): Topology =
+    if (!shouldEmitKeyValues(node)) {
```

Comment:
very minor, may be handle the positive case first i.e. if (shouldEmitKeyValues(node)), negation is slightly harder to read. Again, very minor.